packages/components/src/components/cells/code.tsx

```diff
@@ -0,0 +1,636 @@
+                <div className="flex items-center gap-1">
+                  <TooltipProvider>
+                    <Tooltip>
+                      <TooltipTrigger asChild>
+                        <Button
+                          variant="icon"
+                          className="w-8 px-0"
+                          size="icon"
+                          onClick={() => {
+                            navigator.clipboard.writeText(props.cell.source);
+                            toast.success('Copied to clipboard.');
+                          }}
+                          tabIndex={1}
+                        >
+                          <CopyIcon size={16} />
+                        </Button>
+                      </TooltipTrigger>
+                      <TooltipContent>Copy to clipboard</TooltipContent>
+                    </Tooltip>
+                  </TooltipProvider>
+                </div>
+              </div>
+            </div>
+          ) : (
+            <Header
+              cell={props.cell}
+              runCell={props.onRunCell}
+              stopCell={props.onStopCell}
+              onDeleteCell={!props.readOnly ? props.onDeleteCell : null}
+              generate={props.onGenerate}
+              cellMode={props.cellMode}
+              setCellMode={props.onChangeCellModeType}
+              prompt={props.prompt}
+              setPrompt={props.onChangePrompt}
+              updateFilename={props.onUpdateFileName}
+              filenameError={props.filenameError}
+              setFilenameError={props.onChangeFilenameError}
+              fullscreen={props.fullscreen}
+              setFullscreen={props.onChangeFullscreen}
+              setShowStdio={props.onChangeShowStdio}
+              onAccept={props.onAccept}
+              onRevert={props.onRevert}
+              formatCell={props.onFormatCell}
+              aiEnabled={props.aiEnabled}
+            />
+          )}
+
+          {!props.readOnly && props.cellMode === 'reviewing' ? (
+            <DiffEditor original={props.cell.source} modified={props.newSource} />
+          ) : (
+            <>
+              <div
+                className={cn(!props.readOnly && props.cellMode !== 'off' && 'opacity-50')}
+                id={props.cell.filename}
+              >
+                {props.readOnly ? (
+                  <CodeEditor cell={props.cell} extensions={[]} codeTheme={props.codeTheme} />
+                ) : (
+                  <CodeEditor
+                    // FIXME: make sure that the changes this should be causing in `props.editorExtensions` actually are working
+                    // readOnly={['generating', 'prompting', 'formatting'].includes(props.cellMode)}
+                    cell={props.cell}
+                    updateCellOnServer={props.updateCellOnServer}
+                    extensions={props.editorExtensions}
+                    codeTheme={props.codeTheme}
+                  />
```

Comment:
This is a place where some more testing could take place.

Because I made changes to where the extensions were set up (mentioned in an earlier comment), that had implications on which were enabled in certain modes and in which the editor was read only. I'm like 95% sure I did it right but having somebody with a strong grasp on exactly the expected behavior make sure the editor renders right in each `cellMode` would be helpful.